docs/common/use-table-of-contents.ts

```diff
@@ -0,0 +1,109 @@
+import { useRouter } from 'next/router';
+import { useCallback, useEffect, useState } from 'react';
+
+type TableOfContentsOptions = {
+  /** The root containing element to search for headings and observe them scrolling */
+  root?: Element;
+  /** The query selector to fetch all headings, defaults to `h2, h3` */
+  selector?: string;
+};
+
+/**
+ * Find the headings within an element to generate a table of contents.
+ * This also uses an intersection observer to detect whenever headings are visible.
+ *
+ * @todo Remove "first flash of unstyled content" by extracting headings from MDX
```

Comment:
This isn't connected to the MDX AST heading extraction but should help this FOUC issue. We do need to dynamically fetch the existing headers because we can't purely rely on the MDX AST. It's possible that React components inside the MDX page render additional headings that should be rendered in the TOC.